File: 排序算法/quick_sort.py
# -*- coding: utf-8 -*-

import random

def quicksort(arr):
    """ 非原地算法 """
    if len(arr) < 2:
        return arr
    else:
        # """ 非原地算法 """
        # pivot = arr[0]
        # less = [i for i in arr[1:] if i <= pivot]
        # greater = [i for i in arr[1:] if i > pivot]

        # return quicksort(less) + [pivot] + quicksort(greater)

        """ 原地算法 """
        return quick_sort(arr, 0, len(arr) - 1)

def quick_sort(arr, low, high):
    """ 随机快排 """
    if low < high:
        index = random.randint(low, high)  # 随机快排
        arr[index], arr[high] = arr[high], arr[index]
        left, right = partition(arr, low, high)
        quick_sort(arr, low, left-1)
        quick_sort(arr, right+1, high)
    return arr


def partition(arr, low, high):
    """ 
    划分数组, 以arr[high]为分界线，将小于arr[high]的元素放到arr左边，大于arr[high] 的元素放到arr右边 
    Returns:
        left: arr[low] - arr[left-1] 存放比 arr[high] 小的元素
        right: arr[right+1] - arr[high] 存放比 arr[high] 大的元素
    """

    left = low
    right = high

    current = low
    num = arr[high]
    while current <= right:
        if arr[current] < num:
            arr[left], arr[current] = arr[current], arr[left]
            left += 1
            current += 1
        elif arr[current] > num:
            arr[right], arr[current] = arr[current], arr[right]
            right -= 1
        else:
            current += 1

    return left, right
            


if __name__== "__main__":
    print(quicksort([10, 4, 6, 7, 96, 8,23]))

File: 排序算法/quick_sort.cc
#include <iostream>

using namespace std;

// 对arr[l...r] 部分进行partition操作
// 返回 p， 使得arr[l...p-1] < arr[p]; arr[p+1...r] > arr[p]

template <typename T> 
int __partition(T arr[], int l , int r){

    // // 随机快排
    // swap( arr[l], arr[rand()%(r-l+1)+l]);
    T v = arr[l];

    // arr[l+1...j] < v; arr[j+1...i] > v
    int j = l;
    for (int i = l+1; i<= r; i++){
        if (arr[i] < v){
            swap(arr[++j], arr[i]);
        }
    }
    swap(arr[l], arr[j]);

    return j;
}

// 对 arr[l...r] 部分进行快速排序
template <typename T> 
void __quickSort(T arr[], int l , int r){
    if (l >= r)
        return;
    
    int p = __partition(arr, l, r);
    __quickSort(arr, l, p-1);
    __quickSort(arr, p+1, r);
}

template <typename T>  
void quickSort(T arr[], int n){
    __quickSort(arr, 0, n-1);
}

int main()
{
    int a[10] = {
        10,
        9,
        8,
        4,
        3,
        2,
        17,
        6,
        5,
    };
    quickSort(a, 10);
    for (int i = 0; i < 10; i++)
    {
        cout << a[i] << " ";
    }
    cout << endl;
}